hydra-node/test/Hydra/NetworkSpec.hs

```diff
@@ -34,61 +34,84 @@ import Test.QuickCheck (
+                [ (port1, hn1, node1received)
+                , (port2, hn2, node2received)
+                , (port3, hn3, node3received)
+                ]
 
   describe "0MQ Network" $
     it "broadcasts messages between 3 connected peers" $ do
-      node1received <- newEmptyTMVarIO
-      node2received <- newEmptyTMVarIO
-      node3received <- newEmptyTMVarIO
-      showLogsOnFailure $ \tracer -> failAfter 10 $
-        withZeroMQNetwork tracer (Host lo 55677) [Host lo 55678, Host lo 55679] (atomically . putTMVar node1received) $ \hn1 ->
-          withZeroMQNetwork tracer (Host lo 55678) [Host lo 55677, Host lo 55679] (atomically . putTMVar node2received) $ \hn2 ->
-            withZeroMQNetwork tracer (Host lo 55679) [Host lo 55677, Host lo 55678] (atomically . putTMVar node3received) $ \hn3 ->
-              concurrently_ (assertBroadcastFrom requestTx hn1 [node2received, node3received]) $
-                concurrently_
-                  (assertBroadcastFrom requestTx hn2 [node1received, node3received])
-                  (assertBroadcastFrom requestTx hn3 [node2received, node1received])
+      node1received <- atomically newTQueue
+      node2received <- atomically newTQueue
+      node3received <- atomically newTQueue
+      showLogsOnFailure $ \tracer -> failAfter 10 $ do
+        [port1, port2, port3] <- fmap fromIntegral <$> randomUnusedTCPPorts 3
+        withZeroMQNetwork tracer (Host lo port1) [(Host lo port2), (Host lo port3)] (atomically . writeTQueue node1received) $ \hn1 ->
+          withZeroMQNetwork tracer (Host lo port2) [(Host lo port1), (Host lo port3)] (atomically . writeTQueue node2received) $ \hn2 ->
+            withZeroMQNetwork tracer (Host lo port3) [(Host lo port1), (Host lo port2)] (atomically . writeTQueue node3received) $ \hn3 -> do
+              assertAllNodesBroadcast
+                [ (port1, hn1, node1received)
+                , (port2, hn2, node2received)
+                , (port3, hn3, node3received)
+                ]
 
   describe "Serialisation" $
     it "can roundtrip CBOR encoding/decoding of HydraMessage" $ property $ prop_canRoundtripCBOREncoding @(HydraMessage SimpleTx)
 
-assertBroadcastFrom :: (Eq a, Show a) => a -> Network IO a -> [TMVar IO a] -> IO ()
-assertBroadcastFrom requestTx network receivers =
-  tryBroadcast `catch` \(_ :: HUnitFailure) -> tryBroadcast
+assertAllNodesBroadcast ::
+  [(PortNumber, Network IO Integer, TQueue IO Integer)] ->
+  Expectation
+assertAllNodesBroadcast networks =
+  parallelBroadcast checkQueues networks
  where
-  tryBroadcast = do
-    broadcast network requestTx
-    forM_ receivers $ \var -> failAfter 1 $ atomically (takeTMVar var) `shouldReturn` requestTx
+  parallelBroadcast :: IO () -> [(PortNumber, Network IO Integer, TQueue IO Integer)] -> IO ()
+  parallelBroadcast check [] = check
+  parallelBroadcast check ((port, network, _) : rest) =
+    withAsync (fromIntegral port `broadcastFrom` network) $ \_ -> parallelBroadcast check rest
```

Comment:
Could: This function is passing `check` around to just have it as "core" to the `withAsync` chain.. an alternative (which reads a bit clearer to me, would be to `race_ (parallelBroadcast networks) checkQueues` above and consequently:
```suggestion
  parallelBroadcast :: [(PortNumber, Network IO Integer, TQueue IO Integer)] -> IO ()
  parallelBroadcast [] = pure ()
  parallelBroadcast [(port, network, _)] = fromIntegral port `broadcastFrom` network
  parallelBroadcast ((port, network, _) : rest) =
    withAsync (fromIntegral port `broadcastFrom` network) $ \_ -> parallelBroadcast rest
```